### DISCUSSION TITLE
Allow config of templating util settings from app directly

### DISCUSSION THREAD

**Discussion by philipbushnell-valtech** [Sitecore/jss#1782] [Ideas]
### Is your suggestion related to a problem? Please describe.

In [sitecore-jss-dev-tools/src/templating/utils.ts](https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-dev-tools/src/templating/utils.ts#L42) there are settings for finer control over how items are retrieved yet these settings are not configurable beyond their internal usages (e.g. in [sitecore-jss-dev-tools/src/templating/components.ts](https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-dev-tools/src/templating/components.ts#L34)).

In our app, we have additional co-located files alongside our components (e.g. for testing) but without being able to change the `fileFormat` setting, it fails our build due to it outputting irrelevant files in the generated `componentBuilder` file (and things such as test globals like `describe` are not being explicitly imported so fail).

### Describe the solution you'd like

It would be massively helpful if these could be configured from the application directly (e.g. via [generateComponentBuilder](https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-dev-tools/src/templating/nextjs/generate-component-builder.ts#L20)). It should be relatively quick to update function params to allow better customisation.

### Is there a way to achieve this functionality with exisiting tools?

I've had to `patch-package` a regex setting change to work around this issue but it would be much better to just configure the settings directly in the app (i.e. in `scripts/generate-component-builder/plugins/component-builder.ts`) but I can't.

### Additional information

Copied over from https://github.com/Sitecore/jss/issues/1775 where I incorrectly raised this.

**Comment by addy-pathania** [comment]
@philipbushnell-valtech sounds like we have been trying to implement something similar to this. I have added these details to our backlog for review.

**Reply by bhj** [reply]
Any progress on this?

**Reply by addy-pathania** [reply]
We haven't been able to prioritize this yet due to other high-priority items we've been focused on shipping. However, I've increased the priority of this issue based on the support it's receiving. In the meantime, if you'd like to contribute, you're welcome to submit a PR to our repository.

**Reply by bhj** [reply]
Thank you. I'm using this package tangentially, but I can at least describe my approach:

On first investigation I saw the plugin system and tried to write a plugin that filtered the generated list of components. However, once `generateComponentBuilder` (from `@sitecore-jss/sitecore-jss-dev-tools/nextjs`) is called it's too late; the file is written.

Then I found the `components.ts` plugin, which seemed like it would work, but turns out it was really for registering _extra_ components and doesn't really interact with `generateComponentBuilder`. I think that could be made clearer.

Overall, perhaps the "enumerate" and "output" steps of `generateComponentBuilder` could be separated so that it's possible to write a plugin to modify the list prior to writing the file.
